File: .agents/skills/fastmail/scripts/tools/email.ts
import { JMAPClient } from '../jmap-client.js';
import { NotFoundError } from '../errors.js';

export interface Email {
  id: string;
  blobId: string;
  threadId: string;
  mailboxIds: Record<string, boolean>;
  subject: string;
  from: { name: string; email: string }[] | null;
  to: { name: string; email: string }[] | null;
  cc: { name: string; email: string }[] | null;
  receivedAt: string;
  preview: string;
  hasAttachment: boolean;
  keywords: Record<string, boolean>;
}

export interface Mailbox {
  id: string;
  name: string;
  role: string | null;
  totalEmails: number;
  unreadEmails: number;
}

export interface EmailThread {
  id: string;
  emails: Email[];
  subject: string;
  participants: { name: string; email: string }[];
  latestDate: string;
  emailCount: number;
}

export class EmailTools {
  constructor(private client: JMAPClient) {}
  
  async getMailboxes(): Promise<Mailbox[]> {
    const accountId = await this.client.getAccountId();
    
    const response = await this.client.call([
      ['Mailbox/get', { accountId, properties: ['id', 'name', 'role', 'totalEmails', 'unreadEmails'] }, 'a'],
    ]);
    
    const result = response.methodResponses[0][1] as { list: Mailbox[] };
    return result.list;
  }
  
  async getInboxId(): Promise<string> {
    const mailboxes = await this.getMailboxes();
    const inbox = mailboxes.find(m => m.role === 'inbox');
    if (!inbox) throw new Error('Inbox not found');
    return inbox.id;
  }
  
  async listEmails(mailboxId?: string, limit = 20): Promise<Email[]> {
    const accountId = await this.client.getAccountId();
    const targetMailbox = mailboxId || await this.getInboxId();
    
    const response = await this.client.call([
      ['Email/query', {
        accountId,
        filter: { inMailbox: targetMailbox },
        sort: [{ property: 'receivedAt', isAscending: false }],
        limit,
      }, 'a'],
      ['Email/get', {
        accountId,
        '#ids': { resultOf: 'a', name: 'Email/query', path: '/ids' },
        properties: ['id', 'blobId', 'threadId', 'mailboxIds', 'subject', 'from', 'to', 'cc', 'receivedAt', 'preview', 'hasAttachment', 'keywords'],
      }, 'b'],
    ]);
    
    const result = response.methodResponses[1][1] as { list: Email[] };
    return result.list;
  }
  
  async getEmail(emailId: string): Promise<Email & { textBody: string; htmlBody: string }> {
    const accountId = await this.client.getAccountId();
    
    const response = await this.client.call([
      ['Email/get', {
        accountId,
        ids: [emailId],
        properties: ['id', 'blobId', 'threadId', 'mailboxIds', 'subject', 'from', 'to', 'cc', 'receivedAt', 'preview', 'hasAttachment', 'keywords', 'textBody', 'htmlBody', 'bodyValues'],
        fetchTextBodyValues: true,
        fetchHTMLBodyValues: true,
      }, 'a'],
    ]);
    
    const result = response.methodResponses[0][1] as { list: (Email & { textBody: string; htmlBody: string })[] };
    if (result.list.length === 0) throw new Error('Email not found');
    return result.list[0];
  }
  
  async searchEmails(query: string, limit = 20): Promise<Email[]> {
    const accountId = await this.client.getAccountId();
    
    const response = await this.client.call([
      ['Email/query', {
        accountId,
        filter: { text: query },
        sort: [{ property: 'receivedAt', isAscending: false }],
        limit,
      }, 'a'],
      ['Email/get', {
        accountId,
        '#ids': { resultOf: 'a', name: 'Email/query', path: '/ids' },
        properties: ['id', 'blobId', 'threadId', 'mailboxIds', 'subject', 'from', 'to', 'cc', 'receivedAt', 'preview', 'hasAttachment', 'keywords'],
      }, 'b'],
    ]);
    
    const result = response.methodResponses[1][1] as { list: Email[] };
    return result.list;
  }
  
  async sendEmail(options: {
    to: { name?: string; email: string }[];
    cc?: { name?: string; email: string }[];
    bcc?: { name?: string; email: string }[];
    subject: string;
    textBody: string;
    htmlBody?: string;
    inReplyTo?: string;
    references?: string[];
  }): Promise<{ emailId: string; submissionId: string }> {
    const accountId = await this.client.getAccountId();
    
    // Get identity for sending
    const identityResponse = await this.client.call([
      ['Identity/get', { accountId }, 'a'],
    ]);
    
    const identities = (identityResponse.methodResponses[0][1] as { list: { id: string; email: string }[] }).list;
    if (identities.length === 0) throw new Error('No identity found');
    const identityId = identities[0].id;
    
    // Create email
    const emailCreate: Record<string, unknown> = {
      'mailboxIds': {}, // Will be set by submission
      'to': options.to,
      'subject': options.subject,
      'bodyValues': {
        'body': { value: options.textBody, isEncodingProblem: false, isTruncated: false },
      },
      'textBody': [{ partId: 'body', type: 'text/plain' }],
    };
    
    if (options.cc) emailCreate['cc'] = options.cc;
    if (options.bcc) emailCreate['bcc'] = options.bcc;
    if (options.inReplyTo) emailCreate['inReplyTo'] = [options.inReplyTo];
    if (options.references) emailCreate['references'] = options.references;
    
    if (options.htmlBody) {
      emailCreate['bodyValues'] = {
        'textBody': { value: options.textBody, isEncodingProblem: false, isTruncated: false },
        'htmlBody': { value: options.htmlBody, isEncodingProblem: false, isTruncated: false },
      };
      emailCreate['textBody'] = [{ partId: 'textBody', type: 'text/plain' }];
      emailCreate['htmlBody'] = [{ partId: 'htmlBody', type: 'text/html' }];
    }
    
    const response = await this.client.call([
      ['Email/set', {
        accountId,
        create: { 'draft': emailCreate },
      }, 'a'],
      ['EmailSubmission/set', {
        accountId,
        create: {
          'send': {
            identityId,
            emailId: '#draft',
          },
        },
        onSuccessUpdateEmail: {
          '#send': {
            'mailboxIds/drafts': null,
            'mailboxIds/sent': true,
            'keywords/$draft': null,
          },
        },
      }, 'b'],
    ]);
    
    const emailResult = response.methodResponses[0][1] as { created?: { draft: { id: string } } };
    const submissionResult = response.methodResponses[1][1] as { created?: { send: { id: string } } };
    
    return {
      emailId: emailResult.created?.draft?.id || '',
      submissionId: submissionResult.created?.send?.id || '',
    };
  }
  
  async moveToFolder(emailId: string, targetMailboxId: string, sourceMailboxId?: string): Promise<void> {
    const accountId = await this.client.getAccountId();
    
    const patch: Record<string, boolean | null> = {
      [`mailboxIds/${targetMailboxId}`]: true,
    };
    
    const mailboxesToRemove = new Set<string>();
    if (sourceMailboxId) {
      mailboxesToRemove.add(sourceMailboxId);
    } else {
      const currentEmail = await this.getEmail(emailId);
      for (const mailboxId of Object.keys(currentEmail.mailboxIds || {})) {
        if (mailboxId !== targetMailboxId) {
          mailboxesToRemove.add(mailboxId);
        }
      }
    }
    for (const mailboxId of mailboxesToRemove) {
      patch[`mailboxIds/${mailboxId}`] = null;
    }
    
    await this.client.call([
      ['Email/set', {
        accountId,
        update: {
          [emailId]: patch,
        },
      }, 'a'],
    ]);
  }
  
  async setKeywords(emailId: string, keywords: Record<string, boolean>): Promise<void> {
    const accountId = await this.client.getAccountId();
    
    const patch: Record<string, boolean | null> = {};
    for (const [keyword, value] of Object.entries(keywords)) {
      patch[`keywords/${keyword}`] = value || null;
    }
    
    await this.client.call([
      ['Email/set', {
        accountId,
        update: {
          [emailId]: patch,
        },
      }, 'a'],
    ]);
  }
  
  async deleteEmail(emailId: string): Promise<void> {
    // Move to trash
    const mailboxes = await this.getMailboxes();
    const trash = mailboxes.find(m => m.role === 'trash');
    if (!trash) throw new Error('Trash folder not found');
    
    await this.moveToFolder(emailId, trash.id);
  }

  async getThread(emailId: string): Promise<EmailThread> {
    const accountId = await this.client.getAccountId();
    
    // First get the email to find its threadId
    const response = await this.client.call([
      ['Email/get', {
        accountId,
        ids: [emailId],
        properties: ['id', 'threadId'],
      }, 'a'],
    ]);
    
    const emailResult = response.methodResponses[0][1] as { list: { id: string; threadId: string }[] };
    if (emailResult.list.length === 0) {
      throw new NotFoundError('email', emailId);
    }
    
    const threadId = emailResult.list[0].threadId;
    
    // Now get all emails in this thread
    const threadResponse = await this.client.call([
      ['Email/query', {
        accountId,
        filter: { inThread: threadId },
        sort: [{ property: 'receivedAt', isAscending: true }], // Oldest first in thread
      }, 'a'],
      ['Email/get', {
        accountId,
        '#ids': { resultOf: 'a', name: 'Email/query', path: '/ids' },
        properties: ['id', 'blobId', 'threadId', 'mailboxIds', 'subject', 'from', 'to', 'cc', 'receivedAt', 'preview', 'hasAttachment', 'keywords'],
      }, 'b'],
    ]);
    
    const emails = (threadResponse.methodResponses[1][1] as { list: Email[] }).list;
    
    if (emails.length === 0) {
      throw new NotFoundError('thread', threadId);
    }
    
    // Extract unique participants
    const participantMap = new Map<string, { name: string; email: string }>();
    for (const email of emails) {
      const addParticipant = (p: { name: string; email: string } | null) => {
        if (p && !participantMap.has(p.email)) {
          participantMap.set(p.email, p);
        }
      };
      
      email.from?.forEach(addParticipant);
      email.to?.forEach(addParticipant);
      email.cc?.forEach(addParticipant);
    }
    
    return {
      id: threadId,
      emails,
      subject: emails[0].subject,
      participants: Array.from(participantMap.values()),
      latestDate: emails[emails.length - 1].receivedAt,
      emailCount: emails.length,
    };
  }

  async bulkMoveToFolder(
    emailIds: string[],
    targetMailboxId: string,
    sourceMailboxId?: string
  ): Promise<{ succeeded: string[]; failed: { id: string; error: string }[] }> {
    const accountId = await this.client.getAccountId();
    
    // Build update object for all emails
    const update: Record<string, Record<string, boolean | null>> = {};
    
    for (const emailId of emailIds) {
      update[emailId] = {
        [`mailboxIds/${targetMailboxId}`]: true,
      };
      
      if (sourceMailboxId) {
        update[emailId][`mailboxIds/${sourceMailboxId}`] = null;
      }
    }
    
    const response = await this.client.call([
      ['Email/set', {
        accountId,
        update,
      }, 'a'],
    ]);
    
    const result = response.methodResponses[0][1] as {
      updated?: Record<string, unknown>;
      notUpdated?: Record<string, { type: string; description: string }>;
    };
    
    const succeeded = Object.keys(result.updated || {});
    const failed = Object.entries(result.notUpdated || {}).map(([id, err]) => ({
      id,
      error: err.description || err.type,
    }));
    
    return { succeeded, failed };
  }

  async bulkSetKeywords(
    emailIds: string[],
    keywords: Record<string, boolean>
  ): Promise<{ succeeded: string[]; failed: { id: string; error: string }[] }> {
    const accountId = await this.client.getAccountId();
    
    // Build patch for keywords
    const patch: Record<string, boolean | null> = {};
    for (const [keyword, value] of Object.entries(keywords)) {
      patch[`keywords/${keyword}`] = value || null;
    }
    
    // Apply to all emails
    const update: Record<string, Record<string, boolean | null>> = {};
    for (const emailId of emailIds) {
      update[emailId] = { ...patch };
    }
    
    const response = await this.client.call([
      ['Email/set', {
        accountId,
        update,
      }, 'a'],
    ]);
    
    const result = response.methodResponses[0][1] as {
      updated?: Record<string, unknown>;
      notUpdated?: Record<string, { type: string; description: string }>;
    };
    
    const succeeded = Object.keys(result.updated || {});
    const failed = Object.entries(result.notUpdated || {}).map(([id, err]) => ({
      id,
      error: err.description || err.type,
    }));
    
    return { succeeded, failed };
  }

  async bulkDeleteEmails(
    emailIds: string[]
  ): Promise<{ succeeded: string[]; failed: { id: string; error: string }[] }> {
    // Get trash folder
    const mailboxes = await this.getMailboxes();
    const trash = mailboxes.find(m => m.role === 'trash');
    if (!trash) {
      throw new NotFoundError('mailbox', 'Trash');
    }
    
    return this.bulkMoveToFolder(emailIds, trash.id);
  }
}

File: skills/fastmail/scripts/tools/email.py
"""Email operations via JMAP."""

from __future__ import annotations

from ..jmap_client import JMAPClient
from ..errors import NotFoundError


_EMAIL_LIST_PROPERTIES = [
    "id", "blobId", "threadId", "mailboxIds", "subject",
    "from", "to", "cc", "receivedAt", "preview",
    "hasAttachment", "keywords",
]


class EmailTools:
    def __init__(self, client: JMAPClient):
        self._client = client

    def get_mailboxes(self) -> list[dict]:
        account_id = self._client.get_account_id()
        resp = self._client.call([
            ("Mailbox/get", {"accountId": account_id, "properties": ["id", "name", "role", "totalEmails", "unreadEmails"]}, "a"),
        ])
        return resp["methodResponses"][0][1]["list"]

    def _get_inbox_id(self) -> str:
        mailboxes = self.get_mailboxes()
        inbox = next((m for m in mailboxes if m.get("role") == "inbox"), None)
        if not inbox:
            raise RuntimeError("Inbox not found")
        return inbox["id"]

    def list_emails(self, mailbox_id: str | None = None, limit: int = 20) -> list[dict]:
        account_id = self._client.get_account_id()
        target_mailbox = mailbox_id or self._get_inbox_id()

        resp = self._client.call([
            ("Email/query", {
                "accountId": account_id,
                "filter": {"inMailbox": target_mailbox},
                "sort": [{"property": "receivedAt", "isAscending": False}],
                "limit": limit,
            }, "a"),
            ("Email/get", {
                "accountId": account_id,
                "#ids": {"resultOf": "a", "name": "Email/query", "path": "/ids"},
                "properties": _EMAIL_LIST_PROPERTIES,
            }, "b"),
        ])
        return resp["methodResponses"][1][1]["list"]

    def get_email(self, email_id: str) -> dict:
        account_id = self._client.get_account_id()
        resp = self._client.call([
            ("Email/get", {
                "accountId": account_id,
                "ids": [email_id],
                "properties": [
                    "id", "blobId", "threadId", "mailboxIds", "subject",
                    "from", "to", "cc", "receivedAt", "preview",
                    "hasAttachment", "keywords", "textBody", "htmlBody", "bodyValues",
                ],
                "fetchTextBodyValues": True,
                "fetchHTMLBodyValues": True,
            }, "a"),
        ])
        result_list = resp["methodResponses"][0][1]["list"]
        if not result_list:
            raise NotFoundError("email", email_id)
        return result_list[0]

    def search_emails(self, query: str, limit: int = 20) -> list[dict]:
        account_id = self._client.get_account_id()
        resp = self._client.call([
            ("Email/query", {
                "accountId": account_id,
                "filter": {"text": query},
                "sort": [{"property": "receivedAt", "isAscending": False}],
                "limit": limit,
            }, "a"),
            ("Email/get", {
                "accountId": account_id,
                "#ids": {"resultOf": "a", "name": "Email/query", "path": "/ids"},
                "properties": _EMAIL_LIST_PROPERTIES,
            }, "b"),
        ])
        return resp["methodResponses"][1][1]["list"]

    def send_email(
        self,
        to: list[dict],
        subject: str,
        text_body: str,
        html_body: str | None = None,
        cc: list[dict] | None = None,
        bcc: list[dict] | None = None,
        in_reply_to: str | None = None,
        references: list[str] | None = None,
    ) -> dict:
        account_id = self._client.get_account_id()

        # Get identity for sending
        identity_resp = self._client.call([
            ("Identity/get", {"accountId": account_id}, "a"),
        ])
        identities = identity_resp["methodResponses"][0][1]["list"]
        if not identities:
            raise RuntimeError("No identity found")
        identity_id = identities[0]["id"]

        email_create: dict = {
            "mailboxIds": {},
            "to": to,
            "subject": subject,
            "bodyValues": {
                "body": {"value": text_body, "isEncodingProblem": False, "isTruncated": False},
            },
            "textBody": [{"partId": "body", "type": "text/plain"}],
        }

        if cc:
            email_create["cc"] = cc
        if bcc:
            email_create["bcc"] = bcc
        if in_reply_to:
            email_create["inReplyTo"] = [in_reply_to]
        if references:
            email_create["references"] = references

        if html_body:
            email_create["bodyValues"] = {
                "textBody": {"value": text_body, "isEncodingProblem": False, "isTruncated": False},
                "htmlBody": {"value": html_body, "isEncodingProblem": False, "isTruncated": False},
            }
            email_create["textBody"] = [{"partId": "textBody", "type": "text/plain"}]
            email_create["htmlBody"] = [{"partId": "htmlBody", "type": "text/html"}]

        resp = self._client.call([
            ("Email/set", {
                "accountId": account_id,
                "create": {"draft": email_create},
            }, "a"),
            ("EmailSubmission/set", {
                "accountId": account_id,
                "create": {
                    "send": {
                        "identityId": identity_id,
                        "emailId": "#draft",
                    },
                },
                "onSuccessUpdateEmail": {
                    "#send": {
                        "mailboxIds/drafts": None,
                        "mailboxIds/sent": True,
                        "keywords/$draft": None,
                    },
                },
            }, "b"),
        ])

        email_result = resp["methodResponses"][0][1]
        submission_result = resp["methodResponses"][1][1]

        return {
            "emailId": (email_result.get("created") or {}).get("draft", {}).get("id", ""),
            "submissionId": (submission_result.get("created") or {}).get("send", {}).get("id", ""),
        }

    def move_to_folder(self, email_id: str, target_mailbox_id: str, source_mailbox_id: str | None = None) -> None:
        account_id = self._client.get_account_id()

        patch: dict = {f"mailboxIds/{target_mailbox_id}": True}

        mailboxes_to_remove: set[str] = set()
        if source_mailbox_id:
            mailboxes_to_remove.add(source_mailbox_id)
        else:
            current_email = self.get_email(email_id)
            for mailbox_id in (current_email.get("mailboxIds") or {}).keys():
                if mailbox_id != target_mailbox_id:
                    mailboxes_to_remove.add(mailbox_id)

        for mailbox_id in mailboxes_to_remove:
            patch[f"mailboxIds/{mailbox_id}"] = None

        self._client.call([
            ("Email/set", {
                "accountId": account_id,
                "update": {email_id: patch},
            }, "a"),
        ])

    def set_keywords(self, email_id: str, keywords: dict[str, bool]) -> None:
        account_id = self._client.get_account_id()

        patch: dict = {}
        for keyword, value in keywords.items():
            patch[f"keywords/{keyword}"] = value if value else None

        self._client.call([
            ("Email/set", {
                "accountId": account_id,
                "update": {email_id: patch},
            }, "a"),
        ])

    def delete_email(self, email_id: str) -> None:
        mailboxes = self.get_mailboxes()
        trash = next((m for m in mailboxes if m.get("role") == "trash"), None)
        if not trash:
            raise RuntimeError("Trash folder not found")
        self.move_to_folder(email_id, trash["id"])

    def get_thread(self, email_id: str) -> dict:
        account_id = self._client.get_account_id()

        # Get threadId from email
        resp = self._client.call([
            ("Email/get", {
                "accountId": account_id,
                "ids": [email_id],
                "properties": ["id", "threadId"],
            }, "a"),
        ])
        email_result = resp["methodResponses"][0][1]["list"]
        if not email_result:
            raise NotFoundError("email", email_id)

        thread_id = email_result[0]["threadId"]

        # Get all emails in thread
        thread_resp = self._client.call([
            ("Email/query", {
                "accountId": account_id,
                "filter": {"inThread": thread_id},
                "sort": [{"property": "receivedAt", "isAscending": True}],
            }, "a"),
            ("Email/get", {
                "accountId": account_id,
                "#ids": {"resultOf": "a", "name": "Email/query", "path": "/ids"},
                "properties": _EMAIL_LIST_PROPERTIES,
            }, "b"),
        ])

        emails = thread_resp["methodResponses"][1][1]["list"]
        if not emails:
            raise NotFoundError("thread", thread_id)

        # Extract unique participants
        participants: dict[str, dict] = {}
        for email in emails:
            for field in ("from", "to", "cc"):
                for p in (email.get(field) or []):
                    if p.get("email") and p["email"] not in participants:
                        participants[p["email"]] = p

        return {
            "id": thread_id,
            "emails": emails,
            "subject": emails[0].get("subject", ""),
            "participants": list(participants.values()),
            "latestDate": emails[-1].get("receivedAt", ""),
            "emailCount": len(emails),
        }

    @staticmethod
    def _extract_bulk_result(result: dict) -> dict:
        succeeded = list((result.get("updated") or {}).keys())
        failed = [
            {"id": eid, "error": err.get("description", err.get("type", "unknown"))}
            for eid, err in (result.get("notUpdated") or {}).items()
        ]
        return {"succeeded": succeeded, "failed": failed}

    def bulk_move_to_folder(
        self,
        email_ids: list[str],
        target_mailbox_id: str,
        source_mailbox_id: str | None = None,
    ) -> dict:
        account_id = self._client.get_account_id()

        update: dict = {}
        for eid in email_ids:
            patch: dict = {f"mailboxIds/{target_mailbox_id}": True}
            if source_mailbox_id:
                patch[f"mailboxIds/{source_mailbox_id}"] = None
            update[eid] = patch

        resp = self._client.call([
            ("Email/set", {"accountId": account_id, "update": update}, "a"),
        ])
        return self._extract_bulk_result(resp["methodResponses"][0][1])

    def bulk_set_keywords(self, email_ids: list[str], keywords: dict[str, bool]) -> dict:
        account_id = self._client.get_account_id()

        patch: dict = {}
        for keyword, value in keywords.items():
            patch[f"keywords/{keyword}"] = value if value else None

        update = {eid: dict(patch) for eid in email_ids}

        resp = self._client.call([
            ("Email/set", {"accountId": account_id, "update": update}, "a"),
        ])
        return self._extract_bulk_result(resp["methodResponses"][0][1])

    def bulk_delete_emails(self, email_ids: list[str]) -> dict:
        mailboxes = self.get_mailboxes()
        trash = next((m for m in mailboxes if m.get("role") == "trash"), None)
        if not trash:
            raise NotFoundError("mailbox", "Trash")
        return self.bulk_move_to_folder(email_ids, trash["id"])
